pandas/_libs/parsing.pyx

```diff
@@ -0,0 +1,682 @@
+"""
+import sys
+import re
+
+from cpython cimport PyString_Check, PyUnicode_Check
+
+from libc.stdlib cimport free
+
+cimport cython
+from cython cimport Py_ssize_t
+
+
+from datetime import datetime
+import time
+
+import numpy as np
+cimport numpy as np
+from numpy cimport int64_t, ndarray
+np.import_array()
+
+# Avoid import from outside _libs
+if sys.version_info.major == 2:
+    string_types = basestring
+    from StringIO import StringIO
+else:
+    string_types = str
+    from io import StringIO
+
+
+# dateutil compat
+from dateutil.tz import (tzoffset,
+                         tzlocal as _dateutil_tzlocal,
+                         tzfile as _dateutil_tzfile,
+                         tzutc as _dateutil_tzutc,
+                         tzstr as _dateutil_tzstr)
+from dateutil.relativedelta import relativedelta
+from dateutil.parser import DEFAULTPARSER
+from dateutil.parser import parse as du_parse
+
+
+class DateParseError(ValueError):
+    pass
+
+_nat_strings = set(['NaT', 'nat', 'NAT', 'nan', 'NaN', 'NAN'])
+
+_DEFAULT_DATETIME = datetime(1, 1, 1).replace(hour=0, minute=0,
+                                              second=0, microsecond=0)
+_MONTHS = ['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL',
+           'AUG', 'SEP', 'OCT', 'NOV', 'DEC']
+_MONTH_NUMBERS = {k: i for i, k in enumerate(_MONTHS)}
+_MONTH_ALIASES = {(k + 1): v for k, v in enumerate(_MONTHS)}
+
+cdef object _TIMEPAT = re.compile(r'^([01]?[0-9]|2[0-3]):([0-5][0-9])')
+
+cdef set _not_datelike_strings = set(['a', 'A', 'm', 'M', 'p', 'P', 't', 'T'])
+
+NAT_SENTINEL = object()
+# This allows us to reference NaT without having to import it
+
+
+def parse_datetime_string(date_string, freq=None, dayfirst=False,
+                          yearfirst=False, **kwargs):
+    """parse datetime string, only returns datetime.
+    Also cares special handling matching time patterns.
+
+    Returns
+    -------
+    datetime
+    """
+
+    cdef:
+        object dt
+
+    if not _does_string_look_like_datetime(date_string):
+        raise ValueError('Given date string not likely a datetime.')
+
+    if _TIMEPAT.match(date_string):
+        # use current datetime as default, not pass _DEFAULT_DATETIME
+        dt = du_parse(date_string, dayfirst=dayfirst,
+                      yearfirst=yearfirst, **kwargs)
+        return dt
+
+    try:
+        dt, _, _ = _parse_dateabbr_string(date_string, _DEFAULT_DATETIME, freq)
+        return dt
+    except DateParseError:
+        raise
+    except ValueError:
+        pass
+
+    try:
+        dt = du_parse(date_string, default=_DEFAULT_DATETIME,
+                      dayfirst=dayfirst, yearfirst=yearfirst, **kwargs)
+    except TypeError:
+        # following may be raised from dateutil
+        # TypeError: 'NoneType' object is not iterable
+        raise ValueError('Given date string not likely a datetime.')
+
+    return dt
+
+
+def parse_time_string(arg, freq=None, dayfirst=None, yearfirst=None):
```

Comment:
Do you know what the difference is between this `parse_time_string` and the `parse_datetime_string` above?  
From a superficial look they seem very similar

Reply:
You're not wrong.

Reply:
If you do not tackle that in this PR, would you like to open an issue for it?

Reply:
Just added this to #17652